doc/musig-reference.py

```diff
@@ -0,0 +1,453 @@
+    k_2 = nonce_hash(rand, aggpk, 2, msg, extra_in)
+    assert k_1 != 0
+    assert k_2 != 0
+    R_1_ = point_mul(G, k_1)
+    R_2_ = point_mul(G, k_2)
+    pubnonce = cbytes(R_1_) + cbytes(R_2_)
+    secnonce = bytes_from_int(k_1) + bytes_from_int(k_2)
+    return secnonce, pubnonce
+
+def nonce_agg(pubnonces: List[bytes]) -> bytes:
+    u = len(pubnonces)
+    aggnonce = b''
+    for i in (1, 2):
+        R_i_ = infinity
+        for j in range(u):
+            R_i_ = point_add(R_i_, pointc(pubnonces[j][(i-1)*33:i*33]))
+        R_i = R_i_ if not is_infinite(R_i_) else G
+        aggnonce += cbytes(R_i)
+    return aggnonce
+
+SessionContext = namedtuple('SessionContext', ['aggnonce', 'pubkeys', 'tweaks', 'is_xonly', 'msg'])
+
+def get_session_values(session_ctx: SessionContext) -> tuple[bytes, List[bytes], bytes]:
+    (aggnonce, pubkeys, tweaks, is_xonly, msg) = session_ctx
+    Q, gacc_v, tacc_v = key_agg_internal(pubkeys, tweaks, is_xonly)
+    b = int_from_bytes(tagged_hash('MuSig/noncecoef', aggnonce + bytes_from_point(Q) + msg)) % n
+    R_1 = pointc(aggnonce[0:33])
+    R_2 = pointc(aggnonce[33:66])
+    R = point_add(R_1, point_mul(R_2, b))
+    assert not is_infinite(R)
+    e = int_from_bytes(tagged_hash('BIP0340/challenge', bytes_from_point(R) + bytes_from_point(Q) + msg)) % n
+    return (Q, gacc_v, tacc_v, b, R, e)
+
+def get_session_key_agg_coeff(session_ctx: SessionContext, P: Point) -> int:
+    (_, pubkeys, _, _, _) = session_ctx
+    return key_agg_coeff(pubkeys, bytes_from_point(P))
+
+def sign(secnonce: bytes, sk: bytes, session_ctx: SessionContext) -> bytes:
+    (Q, gacc_v, _, b, R, e) = get_session_values(session_ctx)
+    k_1_ = int_from_bytes(secnonce[0:32])
+    k_2_ = int_from_bytes(secnonce[32:64])
+    assert 0 < k_1_ < n
+    assert 0 < k_2_ < n
+    k_1 = k_1_ if has_even_y(R) else n - k_1_
+    k_2 = k_2_ if has_even_y(R) else n - k_2_
+    d_ = int_from_bytes(sk)
+    assert 0 < d_ < n
+    P = point_mul(G, d_)
+    a = get_session_key_agg_coeff(session_ctx, P)
+    gp = 1 if has_even_y(P) else n - 1
+    g_v = 1 if has_even_y(Q) else n - 1
+    d = g_v * gacc_v * gp * d_ % n
+    s = (k_1 + b * k_2 + e * a * d) % n
+    psig = bytes_from_int(s)
+    pubnonce = cbytes(point_mul(G, k_1_)) + cbytes(point_mul(G, k_2_))
+    assert partial_sig_verify_internal(psig, pubnonce, bytes_from_point(P), session_ctx)
+    return psig
+
+def partial_sig_verify(psig: bytes, pubnonces: List[bytes], pubkeys: List[bytes], tweaks: List[bytes], is_xonly: List[bool], msg: bytes, i: int) -> bool:
+    aggnonce = nonce_agg(pubnonces)
+    session_ctx = SessionContext(aggnonce, pubkeys, tweaks, is_xonly, msg)
+    return partial_sig_verify_internal(psig, pubnonces[i], pubkeys[i], session_ctx)
+
+def partial_sig_verify_internal(psig: bytes, pubnonce: bytes, pk_: bytes, session_ctx: SessionContext) -> bool:
+    (Q, gacc_v, _, b, R, e) = get_session_values(session_ctx)
+    s = int_from_bytes(psig)
+    assert s < n
```

Comment:
Isn't it better to just return `False` (similar to BIP-340)? In general there are a few assertions in this code (keyagg, partial sig agg) that would be triggerable by an attacker. I think that ideally we would assert only for function contracts or negligible edge cases. But we can do that later.

Reply:
Yes, I agree.

I added a fix that uses `assert` consistently throughout the entire codebase (every instance of "fail if" in the spec corresponds to `assert` in the code). Ironically, this individual step is doing the opposite of what you're suggesting 🙃 

But in a future PR, we can replace all of the asserts.